lib/sources/astroport/index.ts

```diff
@@ -0,0 +1,208 @@
+      client,
+      lpContract,
+      height,
+      {
+        total_supply_at: {
+          block: height,
+        },
+      },
+    );
+
+    const path = '/cosmos.bank.v1beta1.Query/SupplyOf';
+    const request = {
+      denom,
+    };
+    const data = QuerySupplyOfRequest.encode(request).finish();
+    const response = await client.abciQuery({ path, data, height });
+    this.logger.trace('Got response %o', response);
+    if (response.code !== 0) {
+      throw new Error(
+        `Tendermint query error: ${response.log} Code: ${response.code}`,
+      );
+    }
+    const assetTotalSupply = QuerySupplyOfResponse.decode(response.value).amount
+      .amount;
+
+    return Number(assetTotalSupply) / Number(lpTotalSupply);
+  };
+
+  getUsersBalances = async (
+    height: number,
+    multipliers: Record<string, number>,
+    cb: CbOnUserBalances,
+  ): Promise<void> => {
+    for (const [assetId, asset] of Object.entries(this.assets)) {
+      const lpContract = await this.getLpContract(height, asset.pair_contract);
+      const exchangeRate = await this.getLpExchangeRate(
+        height,
+        asset.denom,
+        lpContract,
+      );
+      const multiplier = multipliers[assetId] * exchangeRate;
+
+      const client = await this.getClient();
+      let startAfter = undefined;
+      while (true) {
+        this.logger.info(`Fetching accounts at key ${startAfter}`);
+        const accountsData: { accounts: string[] } =
+          await queryContractOnHeight(client, lpContract, height, {
+            all_accounts: {
+              limit: this.paginationLimit,
+              start_after: startAfter,
+            },
+          });
+        const accounts = accountsData.accounts;
+        if (accounts.length == 0) {
+          break;
+        }
+        startAfter = accounts[accounts.length - 1];
+
+        const withConcurrencyLimit = pLimit(this.concurrencyLimit);
+        let balances = (
+          await Promise.allSettled(
+            accounts.map((account) =>
+              withConcurrencyLimit(
+                async () =>
+                  await this.getUserBalance(
+                    account,
+                    lpContract,
+                    height,
+                    assetId,
+                  ),
+              ),
+            ),
+          )
+        ).reduce(
+          (
+            filteredResult: {
+              address: string;
+              balance: string;
+              asset: string;
+            }[],
+            settledResult,
+          ) => {
+            if (settledResult.status === 'fulfilled' && settledResult.value) {
+              filteredResult.push(settledResult.value);
+            }
+            return filteredResult;
+          },
+          [],
+        );
+        balances = balances.filter((balance) => balance.balance != '0');
+        if (balances.length == 0) {
+          continue;
+        }
+
+        balances = balances.map((balance) => {
```

Comment:
You can pass multiplier to `this.getUserBalance` and calculate summary balance there instead of `map` here